MicroProfile Config 2.0 draft 2 (#1185)
OpenLiberty/blogs: posts/2021-03-13-microprofile-config-2.0.adoc

link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html[MicroProfile Config 2.0] is a major version release jam-packed with new features to help make configuring your microservices a breeze. Released as part of link:https://download.eclipse.org/microprofile/microprofile-4.0.1/microprofile-spec-4.0.1.html#microprofile4.0[MicroProfile 4.0], Config 2.0 introduces:

- <<Config-Profile, Config Profile>>- A special property, called `mp.config.profile`, can be used to determine which set of config property values are used.

> - <<ConfigValue, ConfigValue>>- An API class which holds a variety of useful information about a specified config property.

Reviewer: Change,
"An API class which holds a variety of useful information about a specified config property."
to,
"An API class that holds a variety of useful information about a specified config property."

> - <<ConfigProperties, @ConfigProperties>>- An annotation which provides a way to retrieve a number of related config property values, with a shared specified prefix, into a CDI bean.

Reviewer: Change,
"An annotation which provides a way to retrieve a number of related config property values, with a shared specified prefix, into a CDI bean."
to,
"An annotation that provides a way to retrieve a number of related config property values, with a shared specified prefix, into a CDI bean."

> - <<Property-Expression, Property Expression>>- A way of embedding config property values inside of other config property values using the `${}` syntax.

Reviewer: Change,
"A way of embedding config property values inside of other config property values using the `${}` syntax."
to,
"A way of embedding config property values inside of other config property values by using the `${}` syntax."

link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html[MicroProfile Config 2.0] is a major version release jam-packed with new features to help make configuring your microservices a breeze. Released as part of link:https://download.eclipse.org/microprofile/microprofile-4.0.1/microprofile-spec-4.0.1.html#microprofile4.0[MicroProfile 4.0], Config 2.0 introduces:

> - <<Config-Profile, Config Profile>>- a property, called `mp.config.profile`, which can be used to determine which set of config property values are used

Reviewer 2: ```suggestion
- <<Config-Profile, Config Profile>> - A special property, called `mp.config.profile`, can be used to determine which set of config property values are used.
```

> - <<ConfigValue, ConfigValue>>- an API class which holds a bunch of useful information about a specified config property

Reviewer 2: ```suggestion
- <<ConfigValue, ConfigValue>> - An API class which holds a variety of useful information about a specified config property.
```

> - <<ConfigProperties, @ConfigProperties>>- an annotation which provides a way to retrieve a number of related config property values, with a shared specified prefix, into a CDI bean.

Reviewer 2: ```suggestion
- <<ConfigProperties, @ConfigProperties>> - An annotation which provides a way to retrieve a number of related config property values, with a shared specified prefix, into a CDI bean.
```

> - <<Property-Expression, Property Expression>>- a way of defining and expanding config property values inside of other config property values using the `${}` syntax.

Reviewer 2: ```suggestion
- <<Property-Expression, Property Expression>> - A way of embedding config property values inside of other config property values using the `${}` syntax.
```

> 1. You’ll need to tell your build where all the awesome new features are defined. 

Reviewer: Change,
"You’ll need to tell your build where all the awesome new features are defined."
to,
"You need to tell your build where all the awesome new features are defined."

+

> The easiest way to do this it to point to the link:https://search.maven.org/artifact/org.eclipse.microprofile/microprofile/4.0.1/pom[MicroProfile 4.0 API] artifact which includes the link:https://search.maven.org/artifact/org.eclipse.microprofile.config/microprofile-config-api/2.0/jar[MicroProfile Config 2.0 API] artifact.

Reviewer: Change,
"The easiest way to do this it to point to the link:https://search.maven.org/artifact/org.eclipse.microprofile/microprofile/4.0.1/pom[MicroProfile 4.0 API] artifact which includes the link:https://search.maven.org/artifact/org.eclipse.microprofile.config/microprofile-config-api/2.0/jar[MicroProfile Config 2.0 API] artifact."
to,
"The easiest way to do this it to point to the link:https://search.maven.org/artifact/org.eclipse.microprofile/microprofile/4.0.1/pom[MicroProfile 4.0 API] artifact that includes the link:https://search.maven.org/artifact/org.eclipse.microprofile.config/microprofile-config-api/2.0/jar[MicroProfile Config 2.0 API] artifact."

Author: This one stumped me for a while... But I think you're correct

Reviewer 2: `The easiest way to do this it to point...` ->
`The easiest way to do this is to point...`

+

> - If you’re using link:https://maven.apache.org/[Maven] add the following dependency to your pom.xml:

Reviewer: Change,
"If you’re using link:https://maven.apache.org/[Maven] add the following dependency to your pom.xml:"
to,
"If you’re using link:https://maven.apache.org/[Maven], add the following dependency to your pom.xml:"

----
+

1. You'll also need to enable the link:https://openliberty.io/docs/latest/reference/feature/feature-overview.html[feature] in Open Liberty. 
+

> You’ll need to enable `mpConfig-2.0` (MicroProfile Config 2.0’s feature name in Open Liberty) in your server.xml in addition to `cdi-2.0` which enables all the cool injection functionally:

Reviewer 2: ```suggestion
This is done by adding `mpConfig-2.0` (MicroProfile Config 2.0’s feature name in Open Liberty) to the feature list in your server.xml. Additionally `cdi-2.0` is recommended to enable all the cool injection functionally:
```

+
.server.xml
[source,xml]
----
<featureManager>
    <feature>mpConfig-2.0</feature>
    <feature>cdi-2.0</feature>
</featureManager>
----
+

> Alternativelly, you could enable the `microProfile-4.0` link:https://openliberty.io/docs/latest/reference/feature/microProfile-4.0.html[convenience feature], which enables `mpConfig-2.0` and `cdi-2.0` and all the other awesome MicroProfile features.

Reviewer 2: ```suggestion
Alternatively, the `microProfile-4.0` link:https://openliberty.io/docs/latest/reference/feature/microProfile-4.0.html[convenience feature] will enable all of the awesome MicroProfile 4.0 features at once, including `mpConfig-2.0` and `cdi-2.0`.
```

.META-INF/microprofile-config.properties
[source]
----
%testing.server.host=example.test.org
%production.server.host=example.prod.org
server.host=example.org
server.port=9080
server.url=http://${server.host}
----

[#Config-Profile]
=== Config Profile

> link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#configprofile[Config Profiles] allow configuration for different environments and development stages while only one of them is active (e.g. dev, test, prod). The Config Profile is specified using the `mp.config.profile` property, which can be set in any ConfigSource. Once it is set, the corresponding config property values associated with the active profile are used. The `mp.config.profile` property can also be set at application start-up, for example if you’re using Maven and the link:https://github.com/OpenLiberty/ci.maven[liberty-maven-plugin] you can start your app in link:https://openliberty.io/docs/latest/development-mode.html[Dev mode] with:

Reviewer 2: ```suggestion
link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#configprofile[Config Profiles] allow configuration for different environments and development stages while only one of them is active (e.g. dev, test, prod). The active Config Profile is specified using the `mp.config.profile` property, which can be set in any ConfigSource. Once it is set, the corresponding config property values associated with the active profile are used. The `mp.config.profile` property can also be set at application start-up, for example if you’re using Maven and the link:https://github.com/OpenLiberty/ci.maven[liberty-maven-plugin] you can start your app in link:https://openliberty.io/docs/latest/development-mode.html[Dev mode] with:
```

Reviewer 2: Not sure if anything needs to be changed but... Setting the property at startup isn't really an alternative method since you're really just setting a value into a builtin ConfigSource

Author: Yeah I agree. I think it's trying to say: "here's an example of where you might want to set the property", rather than "alternatively to setting it in your application, you can set it here". I can update this if it isn't clear?

[source]
----
mvn liberty:dev -Dliberty.var.mp.config.profile="testing"
----

> Where the `-Dliberty.var.mp.config.profile` argument sets a link:https://maven.apache.org/pom.html#Properties[Maven property] which the liberty-maven-plugin link:https://github.com/scottkurz/ci.maven/blob/f3920800351b6d2c26e62a19008b68093afa48ea/docs/common-server-parameters.md#setting-liberty-configuration-with-maven-project-properties[uses to configure the Liberty app]. In this case, the `mp.config.profile` property is set to `testing`.

Reviewer 2: ```suggestion
The `-Dliberty.var.mp.config.profile` argument sets a link:https://maven.apache.org/pom.html#Properties[Maven property] which the liberty-maven-plugin link:https://github.com/scottkurz/ci.maven/blob/f3920800351b6d2c26e62a19008b68093afa48ea/docs/common-server-parameters.md#setting-liberty-configuration-with-maven-project-properties[uses to configure the Liberty server]. In this case, the `mp.config.profile` property is set to `testing`.
```

- link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#_on_property_level[On property level]: config property names can be set in the following format so that they are used for specific selected profiles: 
+
 %<mp.config.profile>.<original property name> 
+
For example, with `mp.config.profile` set to `testing`, retrieving the config value for "server.host" would use the config property `%testing.server.host` from the <<example-ConfigSource, example ConfigSource>> rather than `server.host`. The value of the property would resolve to `example.test.org`.
+
Similarly, if `mp.config.profile` was set to `production`, retrieving "server.host" would resolve to `example.prod.org`. If `mp.config.profile` was **not** set, retrieving "server.host" would resolve to `example.org`.

- link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#_on_config_source_level[On ConfigSource level]: multiple microprofile-config.properties files can be provided in the following format so they can be used for specific selected profiles:
+
 microprofile-config-<mp.config.profile>.properties 
+

> For example, if a file called microprofile-config-testing.properties was provided on the classpath, with `mp.config.profile` set to `testing` the file would be loaded "on top of" the default microprofile-config.properties file. The config property values from microprofile-config-testing.properties would take precedence.

Reviewer: Change,
"The config property values from microprofile-config-testing.properties would take precedence."
to,
"The config property values from microprofile-config-testing.properties take precedence."

Author: I'm not sure about this one. The behaviour is dependent on the condition from the previous sentence.

> With Config Profiles, your microservice reacts appropriately based on the project stage without changing **any** code and without updating a bunch of config values manually.

Reviewer 2: ```suggestion
With Config Profiles, your microservice is configured appropriately based on the project stage without changing **any** code or needing to update a bunch of config values manually.
```

[source,java]
----
serverNameConfigValue.getSourceName(); // Returns: PropertiesConfigSource[source=file:/<path-to-file>/META-INF/microprofile-config.properties]

> serverNameConfigValue.getSourceOrnial(); // Returns: 100 (the default ordinal value for META-INF/microprofile-config.properties)

Reviewer 2: ```suggestion
serverNameConfigValue.getSourceOrdinal(); // Returns: 100 (the default ordinal value for META-INF/microprofile-config.properties)
```

Author: Wow, good spot

----

[#Property-Expression]
=== Property Expression
Property Expressions provide a way to set and expand variables in property values using the `${}` syntax. For example, the config property `server.url` defined in the <<example-ConfigSource, example ConfigSource>>  as `\http://${server.host}` will be resolved to `\http://example.org` since `server.host` is defined as `example.org`:

[source,java]
----
@Inject
@ConfigProperty(name="server.url")
String url; // Returns: http://example.org (or http://example.test.org if mp.config.profile is set to “testing”)
----

> You can also do some funky expressions such as defining default values, composed expressions, and multiple expressions. link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#property-expressions[The spec] covers these really well.

Reviewer 2: ```suggestion
You can also implement some funky expressions such as defining default values, composed expressions, and multiple expressions. link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#property-expressions[The spec] covers these really well.
```

> Previous working configurations might behave differently if the configuration in use contains values with Property Expressions syntax.

Reviewer 2: ```suggestion
Note: Previously working configurations might now behave differently if the configuration happens to contain values with the Property Expressions syntax (`${}`) in them.
```

[#Extra-Info]
== Some Extra Info For The Return Journey

> For the following examples, we'll use a slightly more rogue example ConfigSource (let's call it "example ConfigSource v2"):

Reviewer: Change,
"For the following examples, we'll use a slightly more rogue example ConfigSource (let's call it "example ConfigSource v2"):"
to,
"For the following examples, we use a slightly more rogue example ConfigSource (let's call it "example ConfigSource v2"):"

Author: Not sure about this one. We use an active voice throughout the rest of the blog: "we'll", "you'll", "let's" etc.

Reviewer: It's about the future tense, which is generally okay for guides, but not other content.

* A value is considered to be "empty" if the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/spi/Converter.html[Converter] being used considers it to be "empty". For example: 

> ** All Converters consider `""`, the empty String, to be empty.

Reviewer 2: Technically the spec does not require there to be a Converter for String (to convert from String to String). So maybe you need to add a bullet that simply says something like "A value is considered to be empty if the raw String is empty and is not converted"? Not sure if I am making this over-complicated?

Author: Hmm I think it would be more clear without the additional comment

** The built-in Converter for String[] considers `","` to be empty (because it is "special"). 

> * From MicroProfile Config 2.0, these "empty" values are no longer valid. Retrieving the values "natively" will now throw a `NoSuchElementException`. E.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:

Reviewer 2: ```suggestion
* From MicroProfile Config 2.0, these "empty" values are no longer valid. Retrieving the values directly using the Config API will now throw a `NoSuchElementException`. E.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:
```

Author: This isn't what I meant by "natively". I meant, retrieving the value without using default values or Optionals:
https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/inject/ConfigProperty.html

But perhaps this should be explained

Author: I've updated this. What'cha think?

> * From MicroProfile Config 2.0, these "empty" values are no longer valid. Retrieving the values "natively" (without defaultValues or Optionals) will now throw a `NoSuchElementException`. E.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:

Reviewer 2: ```suggestion
* From MicroProfile Config 2.0 onwards, these "empty" values are no longer valid. Retrieving the values without using an Optional or a default value will now throw a `NoSuchElementException`. E.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:
```

Author: I think we should probably abort "or a default value" since defining a `defaultValue` is only used if the value "[does not exist](https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/inject/ConfigProperty.html#defaultValue--)", which is _(I think)_ different to "empty".

For example, with `empty.string.array=,` defined in a ConfigSource, 
```
    @Inject
    @ConfigProperty(name = "empty.string.array", defaultValue = "a default value")
    Optional<String[]> emptyStringArray;
```
`emptyStringArray` still returns `Optional.empty`

Similarly, 
```
    @Inject
    @ConfigProperty(name = "empty.string.array", defaultValue = "a default value")
    String[] emptyStringArray;
```
Still throws a `DeploymentException` (caused by a `NoSuchElementException`)

Reviewer 2: "empty" is now the same as "does not exist"

Reviewer 2: are you sure about that second example? providing a defaultValue should mean it does NOT throw an exception

Author: From the conclusion of [this issue](https://github.com/eclipse/microprofile-config/issues/608):

> throw NoSuchElementException when converter returns null. DefaultValue is only a lowest config source.

where returning null is being interpreted as "empty"

Author: The only related TCK I can find is [EmptyValuesTest.java](https://github.com/eclipse/microprofile-config/blob/master/tck/src/main/java/org/eclipse/microprofile/config/tck/emptyvalue/EmptyValuesTest.java) which I think is just testing that the Injection causes a `DeploymentException` which I think would occur for both of our understandings since the `defaultValue` is defined as `""`.

Reviewer 2: yes, defaultValue="" would be the same as not having a default value
but, defaultValue="anything" should never result in a NoSuchElementException (for type `String[]` ... some weird custom `Converter<MyType>` could still return a null)

Author: What makes you say adding "defaultValue="anything" should never result in a NoSuchElementException"? I agree this sounds logical, but I don't think this is what the spec says, and [this discussion](https://github.com/eclipse/microprofile-config/issues/608)

Reviewer 2: array is a built in converter which first splits on ",". String is an automatic converter. For the input "anything", the output would be ["anything"] (non-null). Why do you think it could be null (and thus result in a NSEE)?

[source,java]
----
Config config = ConfigProvider.getConfig();

> config.getValue("empty.property", String.class); // Will throw a `NoSuchElementException`

Reviewer: Change,
"config.getValue("empty.property", String.class); // Will throw a `NoSuchElementException`"
to,
"config.getValue("empty.property", String.class); // throws a `NoSuchElementException`"

> config.getValue("empty.array.property", String[].class); // Will throw a `NoSuchElementException`

Reviewer: Change,
"config.getValue("empty.array.property", String[].class); // Will throw a `NoSuchElementException`"
to,
"config.getValue("empty.array.property", String[].class); // throws a `NoSuchElementException`"

----
@Inject
@ConfigProperty(name = "server.port")
OptionalInt optionalServerPort; // Returns: OptionalInt[9080]
----

[#Incompatibility-changes]
== Heads Up! Incompatibility Changes
If you move up from MicroProfile Config 1.x to 2.0, please take care of the following incompatible changes:

* link:https://javadoc.io/static/org.eclipse.microprofile/microprofile/4.0.1/org/eclipse/microprofile/config/Config.html#getPropertyNames[ConfigSource.getPropertyNames()] is no longer a `default` method. Any implementations of a ConfigSource must implement this method.

> * Previous versions of MP Config would not evaluate Property Expressions. As such, previous working configuration may behave differently (if the configuration contains values with Property Expressions syntax, e.g. `${var.name}`). Property Expressions can be disabled by setting the property `mp.config.property.expressions.enabled` with the value of `false`.

Reviewer: Change,
"Previous versions of MP Config would not evaluate Property Expressions. As such, previous working configuration may behave differently (if the configuration contains values with Property Expressions syntax, e.g. `${var.name}`)."
to,
"Previous versions of MP Config do'ot evaluate Property Expressions. As such, previous working configuration might behave differently (if the configuration contains values with Property Expressions syntax, e.g. `${var.name}`)."
